linkcheck/better_exchook2.py

```diff
@@ -40,201 +40,267 @@
+            output('  File "%s", line %d, in %s' % (filename, lineno, name), out=out)
+            if not os.path.isfile(filename):
+                altfn = fallback_findfile(filename)
+                if altfn:
+                    output(
+                        "    -- couldn't find file, trying this instead: " + altfn,
+                        out=out,
+                    )
+                    filename = altfn
+            linecache.checkcache(filename)
+            line = linecache.getline(filename, lineno, f.f_globals)
+            if line:
+                line = line.strip()
+                output('    line: ' + line, out=out)
+                output('    locals:', out=out)
+                alreadyPrintedLocals = set()
+                for tokenstr in grep_full_py_identifiers(parse_py_statement(line)):
+                    splittedtoken = tuple(tokenstr.split("."))
+                    for token in map(
+                        lambda i: splittedtoken[0:i], range(1, len(splittedtoken) + 1)
+                    ):
+                        if token in alreadyPrintedLocals:
+                            continue
+                        tokenvalue = None
+                        tokenvalue = _trySet(tokenvalue, "<local> ", lambda: pretty_print(_resolveIdentifier(f.f_locals, token)))
+                        tokenvalue = _trySet(tokenvalue, "<global> ", lambda: pretty_print(_resolveIdentifier(f.f_globals, token)))
+                        tokenvalue = _trySet(tokenvalue, "<builtin> ", lambda: pretty_print(_resolveIdentifier(f.f_builtins, token)))
+                        tokenvalue = tokenvalue or "<not found>"
+                        output('      ' + ".".join(token) + " = " + tokenvalue, out=out)
+                        alreadyPrintedLocals.add(token)
+                if len(alreadyPrintedLocals) == 0:
+                    output("       no locals", out=out)
+            else:
+                output('    -- code not available --', out=out)
+            _tb = _tb.tb_next
+            n += 1
+
+    except Exception:
+        output("ERROR: cannot get more detailed exception info because:", out=out)
+        import traceback
+
+        for line in traceback.format_exc().split("\n"):
+            output("   " + line, out=out)
+        output("simple traceback:", out=out)
+        traceback.print_tb(tb, None, out)
+
+    import types
+
+    def _some_str(value):
+        try:
+            return str(value)
+        except Exception:
+            return '<unprintable %s object>' % type(value).__name__
+
+    def _format_final_exc_line(etype, value):
+        valuestr = _some_str(value)
+        if value is None or not valuestr:
+            line = "%s" % etype
+        else:
+            line = "%s: %s" % (etype, valuestr)
+        return line
+
+    if (
+        isinstance(etype, BaseException)
+        or (hasattr(types, "InstanceType") and isinstance(etype, types.InstanceType))
+        or etype is None
+        or type(etype) is str
```

Comment:
string exceptions were banned in Python 2.something.  Also, I think for string exceptions etype would be `str`, not an instance of str?

Can etype ever be None?

Can etype ever be an _instance_ of BaseException, rather than a _subclass_ of BaseException?

What is this code doing.